File: testSuperResolutionGanModel.R
library( keras )
library( ANTsRNet )
library( magick )
library( ANTsR )
library( tensorflow )

tf$compat$v1$disable_eager_execution()

keras::backend()$clear_session()

dataset <- 'vangogh2photo'

dataDirectory <- paste0( getwd(), "/", dataset )
if( ! dir.exists( dataDirectory ) )
  {
  zippedFile <- tempfile( pattern = dataset, fileext = ".zip" )

  if( ! file.exists( zippedFile ) )
    {
    download.file(
      url = paste0( "https://people.eecs.berkeley.edu/~taesung_park/CycleGAN/datasets/", dataset, ".zip" ),
      destfile = zippedFile )
    unzip( zippedFile )
    }
  }


scaleFactor <- 4
highResolutionImageSize <- c( 256, 256, 3 )
lowResolutionImageSize <- c( as.integer( highResolutionImageSize[1] / scaleFactor ),
                             as.integer( highResolutionImageSize[2] / scaleFactor ),
                             3 )

ganModel <- SuperResolutionGanModel$new(
   lowResolutionImageSize = lowResolutionImageSize,
   scaleFactor = scaleFactor, useImageNetWeights = TRUE )

cat( "Reading training set A files.\n" )
trainingFilesA <- list.files( paste0( dataDirectory, '/trainA/' ),
  pattern = ".jpg", full.names = TRUE )
numberOfTrainingAFiles <- length( trainingFilesA )

X_trainLowResolution <- array( data = 0, dim = c( numberOfTrainingAFiles, lowResolutionImageSize ) )
X_trainHighResolution <- array( data = 0, dim = c( numberOfTrainingAFiles, highResolutionImageSize ) )

pb <- txtProgressBar( min = 0, max = length( trainingFilesA ), style = 3 )
for( i in seq_len( numberOfTrainingAFiles ) )
  {
  # high resolution
  image <- image_read( trainingFilesA[i] )
  image <- image_data( image_scale( image, "256x256" ) )
  image <- as.double( aperm( image, c( 3, 2, 1 ) ) )
  image <- image / 127.5 - 1.0
  X_trainHighResolution[i,,,] <- as.double( image )

  # low resolution
  image <- image_read( trainingFilesA[i] )
  image <- image_data( image_scale( image, "64x64" ) )
  image <- as.double( aperm( image, c( 3, 2, 1 ) ) )
  image <- image / 127.5 - 1.0
  X_trainLowResolution[i,,,] <- as.double( image )

  setTxtProgressBar( pb, i )
  }
cat( "\nDone.\n\n" )


ganModel$train( X_trainLowResolution, X_trainHighResolution,
  numberOfEpochs = 200, batchSize = 4, sampleInterval = 25,
  sampleFilePrefix = "./SuperResolutionGanSampleImages/sample" )

File: testSuperResolutionGanModel.py
import requests
import tempfile
import numpy as np
import os
import cv2
import zipfile
import glob

import keras.backend as K

import ants
import antspynet

K.clear_session()

dataset = 'vangogh2photo'

data_directory = os.getcwd() + "/" + dataset
if not os.path.exists(data_directory):
    zipped_file = tempfile.NamedTemporaryFile(suffix=".zip", dir=os.getcwd())
    url = "https://people.eecs.berkeley.edu/~taesung_park/CycleGAN/datasets/" + dataset + ".zip"
    r = requests.get(url)
    with open(zipped_file.name, 'wb') as f:
        f.write(r.content)

    with zipfile.ZipFile(zipped_file.name, 'r') as zip_ref:
        zip_ref.extractall(os.getcwd())

scale_factor = 4
high_resolution_image_size = (256, 256, 3)
low_resolution_image_size = (int(high_resolution_image_size[0] / scale_factor),
                             int(high_resolution_image_size[1] / scale_factor),
                             3 )

gan_model = antspynet.SuperResolutionGanModel(
   low_resolution_image_size = low_resolution_image_size,
   scale_factor = scale_factor )

print("Reading training set A files.")
trainingA_files = glob.glob(data_directory + "/trainA/*.jpg")
number_of_trainingA_files = len(trainingA_files)
X_train_low_resolution = np.zeros(shape=(number_of_trainingA_files, *low_resolution_image_size))
X_train_high_resolution = np.zeros(shape=(number_of_trainingA_files, *high_resolution_image_size))

for i in range(number_of_trainingA_files):
   image = cv2.imread(trainingA_files[i])
   image = cv2.resize(image, high_resolution_image_size[:2])
   image = image / 127.5 - 1.0
   X_train_high_resolution[i,:,:,:] = image

   image = cv2.resize(image, low_resolution_image_size[:2])
   X_train_low_resolution[i,:,:,:] = image

print("Done.")

gan_model.train(X_train_low_resolution, X_train_high_resolution,
  number_of_epochs=200, batch_size=4, sample_interval=25,
  sample_file_prefix="./SuperResolutionGanSampleImages_Py/sample" )

